object literal defined using := panics with "corrupt object"

## Expected Behavior

Running this package through `opa run demo.rego`,
```
package demo

foo[f] {
	k = "key"
	v := "val"
	f := { k: v }
}
```
I get this:
```interactive
$ opa run demo.rego
OPA 0.10.3-dev (commit a9f819f0-dirty, built at 2018-12-20T13:03:31Z)

Run 'help' to see a list of commands.

> data.demo.foo
[
  {
    "key": "val"
  }
]
>
```

Now, I'd expect the same output from the package
```
package demo

foo[f] {
	k := "key" # <<< 
	v := "val"
	f := { k: v }
}
```

### Actual Behavior

```interactive
$ opa run demo.rego
panic: corrupt object [recovered]
        panic: corrupt object

goroutine 1 [running]:
github.com/open-policy-agent/opa/ast.(*Compiler).compile.func1()
        /Users/stephan/go/src/github.com/open-policy-agent/opa/ast/compile.go:611 +0x6a
panic(0x4670f60, 0x482e9f0)
        /usr/local/Cellar/go/1.11.3/libexec/src/runtime/panic.go:513 +0x1b9
github.com/open-policy-agent/opa/ast.(*object).Iter(0xc0001cd1d0, 0xc000224d20, 0x20, 0x46d77c0)
        /Users/stephan/go/src/github.com/open-policy-agent/opa/ast/term.go:1563 +0xf5
github.com/open-policy-agent/opa/ast.(*object).Foreach(0xc0001cd1d0, 0xc0001d0aa0)
        /Users/stephan/go/src/github.com/open-policy-agent/opa/ast/term.go:1586 +0x53
github.com/open-policy-agent/opa/ast.WalkBeforeAndAfter(0x483d140, 0xc0001cef60, 0x4753660, 0xc0001cd1d0)
        /Users/stephan/go/src/github.com/open-policy-agent/opa/ast/visit.go:104 +0x887
github.com/open-policy-agent/opa/ast.Walk(0x4833420, 0xc0000c84d8, 0x4753660, 0xc0001cd1d0)
        /Users/stephan/go/src/github.com/open-policy-agent/opa/ast/visit.go:30 +0x7b
github.com/open-policy-agent/opa/ast.WalkBeforeAndAfter(0x483d140, 0xc0001cef50, 0x4735a60, 0xc0001d0440)
        /Users/stephan/go/src/github.com/open-policy-agent/opa/ast/visit.go:98 +0xe64
github.com/open-policy-agent/opa/ast.Walk(0x4833420, 0xc0000c84d8, 0x4735a60, 0xc0001d0440)
        /Users/stephan/go/src/github.com/open-policy-agent/opa/ast/visit.go:30 +0x7b
github.com/open-policy-agent/opa/ast.WalkBeforeAndAfter(0x483d140, 0xc0001ceee0, 0x47541c0, 0xc0002260a0)
        /Users/stephan/go/src/github.com/open-policy-agent/opa/ast/visit.go:86 +0x9d7
github.com/open-policy-agent/opa/ast.Walk(0x4833420, 0xc0000c84d8, 0x47541c0, 0xc0002260a0)
        /Users/stephan/go/src/github.com/open-policy-agent/opa/ast/visit.go:30 +0x7b
github.com/open-policy-agent/opa/ast.WalkBeforeAndAfter(0x483d140, 0xc0001cedb0, 0x47367e0, 0xc0001d0a80)
        /Users/stephan/go/src/github.com/open-policy-agent/opa/ast/visit.go:76 +0xb8f
github.com/open-policy-agent/opa/ast.Walk(0x4833420, 0xc0000c84d8, 0x47367e0, 0xc0001d0a80)
        /Users/stephan/go/src/github.com/open-policy-agent/opa/ast/visit.go:30 +0x7b
github.com/open-policy-agent/opa/ast.WalkBeforeAndAfter(0x483d140, 0xc0001cea80, 0x471f040, 0xc0001d55c0)
        /Users/stephan/go/src/github.com/open-policy-agent/opa/ast/visit.go:61 +0xeec
github.com/open-policy-agent/opa/ast.Walk(0x4833420, 0xc0000c84d8, 0x471f040, 0xc0001d55c0)
        /Users/stephan/go/src/github.com/open-policy-agent/opa/ast/visit.go:30 +0x7b
github.com/open-policy-agent/opa/ast.WalkBeforeAndAfter(0x483d140, 0xc0001cea10, 0x46f22a0, 0xc0000d9e50)
        /Users/stephan/go/src/github.com/open-policy-agent/opa/ast/visit.go:49 +0x28d
github.com/open-policy-agent/opa/ast.Walk(0x4833420, 0xc0000c84d8, 0x46f22a0, 0xc0000d9e50)
        /Users/stephan/go/src/github.com/open-policy-agent/opa/ast/visit.go:30 +0x7b
github.com/open-policy-agent/opa/ast.WalkRules(0x46f22a0, 0xc0000d9e50, 0xc0001d05c0)
        /Users/stephan/go/src/github.com/open-policy-agent/opa/ast/visit.go:227 +0x9d
github.com/open-policy-agent/opa/ast.(*Compiler).rewriteLocalAssignments(0xc0000e1dd0)
        /Users/stephan/go/src/github.com/open-policy-agent/opa/ast/compile.go:798 +0x10d
github.com/open-policy-agent/opa/ast.(*Compiler).rewriteLocalAssignments-fm()
        /Users/stephan/go/src/github.com/open-policy-agent/opa/ast/compile.go:205 +0x2a
github.com/open-policy-agent/opa/ast.(*Compiler).compile(0xc0000e1dd0)
        /Users/stephan/go/src/github.com/open-policy-agent/opa/ast/compile.go:616 +0x81
github.com/open-policy-agent/opa/ast.(*Compiler).Compile(0xc0000e1dd0, 0xc000225838)
        /Users/stephan/go/src/github.com/open-policy-agent/opa/ast/compile.go:249 +0x274
github.com/open-policy-agent/opa/runtime.compileAndStoreInputs(0x483da40, 0xc0000ba008, 0x48428e0, 0xc0000d9e00, 0x48336c0, 0xc0001c2c90, 0xc000184480, 0xa, 0x0, 0x46a0440)
        /Users/stephan/go/src/github.com/open-policy-agent/opa/runtime/runtime.go:411 +0x1b7
github.com/open-policy-agent/opa/runtime.NewRuntime(0x483da40, 0xc0000ba008, 0xc0000d6240, 0x24, 0xc00017a220, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/stephan/go/src/github.com/open-policy-agent/opa/runtime/runtime.go:168 +0x2ff
github.com/open-policy-agent/opa/cmd.init.7.func1(0xc000164a00, 0xc00016c2d0, 0x1, 0x1)
        /Users/stephan/go/src/github.com/open-policy-agent/opa/cmd/run.go:113 +0x42b
github.com/open-policy-agent/opa/vendor/github.com/spf13/cobra.(*Command).execute(0xc000164a00, 0xc00016c2a0, 0x1, 0x1, 0xc000164a00, 0xc00016c2a0)
        /Users/stephan/go/src/github.com/open-policy-agent/opa/vendor/github.com/spf13/cobra/command.go:766 +0x2cc
github.com/open-policy-agent/opa/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x4d06fe0, 0xc0000fdf78, 0x45e69bd, 0x477c3cf)
        /Users/stephan/go/src/github.com/open-policy-agent/opa/vendor/github.com/spf13/cobra/command.go:852 +0x2fd
github.com/open-policy-agent/opa/vendor/github.com/spf13/cobra.(*Command).Execute(0x4d06fe0, 0x0, 0xc00013de60)
        /Users/stephan/go/src/github.com/open-policy-agent/opa/vendor/github.com/spf13/cobra/command.go:800 +0x2b
main.main()
        /Users/stephan/go/src/github.com/open-policy-agent/opa/main.go:12 +0x2d
```

--- a/ast/compile.go
+++ b/ast/compile.go
@@ -841,15 +841,32 @@ func (c *Compiler) rewriteLocalAssignments() {
 
 			// Rewrite vars in head that refer to locally declared vars in the body.
 			vis := NewGenericVisitor(func(x interface{}) bool {
-				switch x := x.(type) {
-				case *Term:
-					if v, ok := x.Value.(Var); ok {
-						if gv, ok := declared[v]; ok {
-							x.Value = gv
-							return true
+
+				term, ok := x.(*Term)
+				if !ok {
+					return false
+				}
+
+				switch v := term.Value.(type) {
+				case Object:
+					// Make a copy of the object because the keys may be mutated.
+					cpy, _ := v.Map(func(k, v *Term) (*Term, *Term, error) {
+						if vark, ok := k.Value.(Var); ok {
+							if gv, ok := declared[vark]; ok {
+								k = k.Copy()
+								k.Value = gv
+							}
 						}
+						return k, v, nil
+					})
+					term.Value = cpy
+				case Var:
+					if gv, ok := declared[v]; ok {
+						term.Value = gv
+						return true
 					}
 				}
+
 				return false
 			})
 
@@ -1029,7 +1046,12 @@ func (qc *queryCompiler) rewriteLocalAssignments(_ *QueryContext, body Body) (Bo
 	}
 	qc.rewritten = make(map[Var]Var, len(declared))
 	for k, v := range declared {
-		qc.rewritten[v] = k
+		// The vars returned during the rewrite will include all seen vars,
+		// even if they're not declared with an assignment operation. We don't
+		// want to include these inside the rewritten set though.
+		if Compare(k, v) != 0 {
+			qc.rewritten[v] = k
+		}
 	}
 	return body, nil
 }
@@ -2554,6 +2576,14 @@ func rewriteDeclaredVarsInTerm(g *localVarGenerator, stack *localDeclaredVars, t
 			return true, errs
 		}
 		return false, errs
+	case Object:
+		cpy, _ := v.Map(func(k, v *Term) (*Term, *Term, error) {
+			kcpy := k.Copy()
+			_, errs = rewriteDeclaredVarsInTerm(g, stack, kcpy, errs)
+			_, errs = rewriteDeclaredVarsInTerm(g, stack, v, errs)
+			return kcpy, v, nil
+		})
+		term.Value = cpy
 	case *ArrayComprehension:
 		errs = rewriteDeclaredVarsInArrayComprehension(g, stack, v, errs)
 	case *SetComprehension:

--- a/ast/compile_test.go
+++ b/ast/compile_test.go
@@ -1106,6 +1106,19 @@ func TestCompilerRewriteLocalAssignments(t *testing.T) {
 	head_array_comprehensions = [[x] | x := 1]
 	head_set_comprehensions = {[x] | x := 1}
 	head_object_comprehensions = {k: [x] | k := "foo"; x := 1}
+
+	rewritten_object_key {
+		k := "foo"
+		{k: 1}
+	}
+
+	rewritten_object_key_head[[{k: 1}]] {
+		k := "foo"
+	}
+
+	rewritten_object_key_head_value = [{k: 1}] {
+		k := "foo"
+	}
 	`)
 
 	c.Modules["test2"] = MustParseModule(`package test
@@ -1163,6 +1176,10 @@ func TestCompilerRewriteLocalAssignments(t *testing.T) {
 	head_array_comprehensions = [[__local21__] | __local21__ = 1]
 	head_set_comprehensions = {[__local22__] | __local22__ = 1}
 	head_object_comprehensions = {__local23__: [__local24__] | __local23__ = "foo"; __local24__ = 1}
+
+	rewritten_object_key = true { __local25__ = "foo"; {__local25__: 1} }
+	rewritten_object_key_head[[{__local26__: 1}]] { __local26__ = "foo" }
+	rewritten_object_key_head_value = [{__local27__: 1}] { __local27__ = "foo" }
 	`)
 
 	if len(module1.Rules) != len(expectedModule.Rules) {
@@ -2209,6 +2226,7 @@ func TestQueryCompilerRewrittenVars(t *testing.T) {
 		vars map[string]string
 	}{
 		{"assign", "a := 1", map[string]string{"__local0__": "a"}},
+		{"suppress only seen", "b = 1; a := b", map[string]string{"__local0__": "a"}},
 	}
 	for _, tc := range tests {
 		t.Run(tc.note, func(t *testing.T) {

